extension/src/time_vector.rs

```diff
@@ -368,6 +368,91 @@ CREATE AGGREGATE rollup(\n\
     ],
 );
 
+#[pg_schema]
+pub mod toolkit_experimental {
+    use super::*;
+
+    #[pg_extern(immutable, parallel_safe)]
+    pub fn asof_join<'a, 'b>(
+        from: Timevector_TSTZ_F64<'a>,
+        into: Timevector_TSTZ_F64<'b>,
+    ) -> TableIterator<
+        'a,
+        (
+            name!(time, crate::raw::TimestampTz),
+            name!(value1, Option<f64>),
+            name!(value2, f64),
+        ),
+    > {
+        let mut from = from
+            .into_iter()
+            .map(|points| (points.ts.into(), points.val))
+            .peekable();
+        let into = into.into_iter().map(|points| (points.ts, points.val));
+        let (mut from_time, mut from_val) = from.next().unwrap();
```

Comment:
The error from `.unwrap()` isn't particularly user friendly:
```suggestion
        let (mut from_time, mut from_val) = from.next().expect("Must have at least one point in joining column");
```

Reply:
Good point, however it's also going to have potentially ugly behavior if the into vector is empty, so I added an assert that neither is empty to the start of this function.